lib/Runtime/Library/InJavascript/Intl.js

```diff
@@ -286,114 +230,258 @@
+            if (availableLocale !== undefined) {
+                result.locale = availableLocale;
+                if (requestedLocales[i] !== parsedLangtag.base) {
+                    result.extension = parsedLangtag.extension;
+                }
+
+                return result;
+            }
         }
 
-        return resolvedLocale;
-    }
+        result.locale = GetDefaultLocale();
+        return result;
+    };
 
-    // ECMA402 section 9.2.5 (#sec-unicodeextensionvalue)
-    const UnicodeExtensionSubtags = function (extensionString) {
-        if (!extensionString) {
-            return [];
-        }
+    const BestFitMatcher = LookupMatcher;
 
-        const extensions = _.filter(_.split(extensionString, "-"), (el) => !!el);
-        let uStart = -1;
-        for (let i = 0; i < extensions.length; i++) {
-            const ex = extensions[i];
-            if (ex === "u" && uStart === -1 && i + 1 < extensions.length) {
-                uStart = i + 1;
-            } else if (ex.length === 1 && uStart !== -1) {
-                // in -u-k1-value1-k2-value2-x-private, we have hit -x, return [k1, value1, k2, value2]
-                return _.slice(extensions, uStart, i);
+    /**
+     * Determine a value for a given key in the given extension string
+     * 
+     * ECMA-402: #sec-unicodeextensionvalue
+     * 
+     * @param {String} extension the full unicode extension, such as "-u-co-phonebk-kf-true"
+     * @param {String} key the specific key we are looking for in the extension, such as "co"
+     */
+    const UnicodeExtensionValue = function (extension, key) {
+        raise.assert(key.length === 2);
+        const size = extension.length;
+
+        // search for the key-value pair
+        let pos = _.stringIndexOf(extension, `-${key}-`);
+        if (pos !== -1) {
+            const start = pos + 4;
+            let end = start;
+            let k = start;
+            let done = false;
+            while (!done) {
+                const e = _.stringIndexOf(extension, "-", k);
+                const len = e === -1 ? size - k : e - k;
+                if (len === 2) {
+                    done = true;
+                } else if (e === -1) {
+                    end = size;
+                    done = true;
+                } else {
+                    end = e;
+                    k = e + 1;
+                }
             }
-        }
 
-        if (uStart === -1) {
-            return [];
-        } else {
-            return _.slice(extensions, uStart);
+            return _.substring(extension, start, end);
         }
-    };
 
-    const resolveLocaleHelper = function (locale, fitter, extensionFilter = []) {
-        const localeParts = platform.builtInRegexMatch(locale, LANG_TAG_RE);
-        if (!localeParts || !localeParts[1]) {
+        // search for the key with no associated value
+        pos = _.stringIndexOf(extension, `-${key}`);
```

Comment:
This might be more of a spec issue, but shouldn't this be checking the last index of `-${key}` and not the first?